Use leader election in globalnet controller

tpantelis opened the pull request:
Due to timing of notifications during gateway node transition, it's been observed that there can be a small window of a few seconds where the controllers on the previously active gateway are still running when the controllers on the newly active gateway are started. This results in two instances of the controllers running simultaneously in the cluster which can cause strange behavior and inconsistent results.

To alleviate this scenario, use leader election in the `GatewayMonitor` to ensure that the instance on the newly active gateway doesn't start its controllers until the previous instance has stopped its controllers and released the leader lock. There's a couple things to be aware of:

 - The K8s leader election functionality periodically renews the lease. If it can't be renewed prior to the `RenewDeadline`, it stops. For our usage, we don't have instances concurrently vying for leadership so we really don't need to keep renewing the lease. Ideally we would set the `RenewDeadline` very high to essentially disable it but it needs to be less than the `LeaseDuration` setting which we don't want too high. So we handle renewal failure by simply re-running leader election.

 - When transitioning to a non-gateway or shutting down, the `GatewayMonitor` cancels leader election which causes it to release the lock, which clears the `HolderIdentity` field and sets `LeaseDuration` to one second. This enables the next instance to quickly acquire the lock. However, if the controller instance crashes and doesn't properly release the lock, the next instance will have to await the `LeaseDuration` period before acquiring the lock. So we don't want `LeaseDuration` set too high and we don't want it too low either to give the current instance enough time to complete stopping its controllers. I picked 20 sec for `LeaseDuration` and 15 sec for `RenewDeadline`. So if the current instance takes longer than 20 sec to stop the controllers, it's still possible there could be two instance running concurrently but this scenario should be highly unlikely.

tpantelis committed:
Use leader election in globalnet controller

Due to timing of notifications during gateway node transition, it's been
observed that there can be a small window of a few seconds where the
controllers on the previously active gateway are still running when the
controllers on the newly active gateway are started. This results in two
instances of the controllers running simultaneously in the cluster which
can cause strange behavior and inconsistent results.

To alleviate this scenario, use leader election in the GatewayMonitor to
ensure that the instance on the newly active gateway doesn't start its
controllers until the previous instance has stopped its controllers and
released the leader lock. There's a couple things to be aware of:

 - The K8s leader election functionality periodically renews the lease.
   If it can't be renewed prior to the RenewDeadline, it stops. For our
   usage, we don't have instances concurrently vying for leadership so
   we really don't need to keep renewing the lease. Ideally we would set
   the RenewDeadline very high to essentially disable it but it needs to
   be less than the LeaseDuration setting which we don't want too high.
   So we handle renewal failure by simply re-running leader election.

 - When transitioning to a non-gateway or shutting down, the GatewayMonitor
   cancels leader election which causes it to release the lock, which
   clears the HolderIdentity field and sets LeaseDuration to one second.
   This enables the next instance to quickly acquire the lock. However,
   if the controller instance crashes and doesn't properly release the
   lock, the next instance will have to await the LeaseDuration period
   before acquiring the lock. So we don't want LeaseDuration set too
   high and we don't want it too low either to give the current
   instance enough time to complete stopping its controllers. I
   picked 20 sec for LeaseDuration and 15 sec for RenewDeadline. So
   if the current instance takes longer than 20 sec to stop the
   controllers, it's still possible there could be two instance running
   concurrently but this scenario should be unlikely.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

skitt commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -39,23 +41,44 @@ import (
+		LeaderElectionConfig:    config.LeaderElectionConfig,
 	}
 
+	if gatewayMonitor.LeaseDuration == 0 {
```
Could you add the explanations for the default values as comments here?

skitt commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -506,6 +603,28 @@ func (g *gatewayMonitor) markRemoteClusterTraffic(remoteCidr string, addRules bo
+	}
+}
+
+func (l *LeaderElectionInfo) awaitStopped(timeout int) {
```
The timeout would be clearer if it was passed directly as a `time.Duration` (and on `stopControllers`).

skitt commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -134,15 +157,17 @@ func (g *gatewayMonitor) Start() error {
 	g.baseController.Stop()
 
-	g.stopControllers()
+	g.stopControllers(10)
```
Please add a comment to explain the value.

skitt commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -353,9 +439,10 @@ func (g *gatewayMonitor) startControllers() error {
 }
 
-func (g *gatewayMonitor) stopControllers() {
+func (g *gatewayMonitor) stopControllers(timeout int) {
```
```suggestion
func (g *gatewayMonitor) stopControllers(leaderElectionTimeout int) {
```

but that’s not great, because the timeout isn’t used for the election timeout itself, but when stopping the controller; so perhaps `leaderElectionControllerTimeout` but that’s a mouthful.

skitt replied:
The timeout handling here might be better done with a context with a timeout ;-). (Just a thought, I haven’t tried it.)

tpantelis replied:
I changed it to a context.

skitt commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -506,6 +603,28 @@ func (g *gatewayMonitor) markRemoteClusterTraffic(remoteCidr string, addRules bo
+		select {
+		case <-l.stopped:
+		case <-time.After(time.Duration(timeout) * time.Second):
+			logger.Warning("Timed out waiting for leader election stopped")
```
```suggestion
			logger.Warning("Timed out waiting for leader election to stop")
```

sridhargaddam commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -252,11 +290,75 @@ func (g *gatewayMonitor) handleRemovedEndpoint(obj runtime.Object, _ int) bool {
+		return nil
+	}
+
+	logger.Infof("Starting controllers")
```
nit: 
```suggestion
	logger.Info("Starting controllers")
```

sridhargaddam reviewed (approved):
Thanks for working on the fix @tpantelis

tpantelis commented:
@skitt @sridhargaddam So if leadership is lost due to failure after the `RenewDeadline` expires, currently we try to re-acquire leadership but we don't stop the controllers. A failure can really only happen if the node loses contact with the API server and can't update the `Lease` resource. I did this b/c we're only using leader election to prevent concurrent instances running during gateway transition. However, it's possible there could be a gateway transition during the time that leadership is lost and a new instance is able to contact the API server to acquire the lock (after the `LeaseDuration` expires). In this case, we'd have two controller instances running although the first is most likely effectively disabled anyway since most likely it can't contact the API server. However, if it regains contact,  before it observes the gateway has transitioned, it could try to process globalnet resources and causes strange inconsistencies like we've seen. This is an edge case but perhaps we should stop the controllers if leadership is lost due to failure, especially since the controllers are likely offline as well.  WDYT?

sridhargaddam commented:
> @skitt @sridhargaddam So if leadership is lost due to failure after the `RenewDeadline` expires, currently we try to re-acquire leadership but we don't stop the controllers. A failure can really only happen if the node loses contact with the API server and can't update the `Lease` resource. I did this b/c we're only using leader election to prevent concurrent instances running during gateway transition. However, it's possible there could be a gateway transition during the time that leadership is lost and a new instance is able to contact the API server to acquire the lock (after the `LeaseDuration` expires). In this case, we'd have two controller instances running although the first is most likely effectively disabled anyway since most likely it can't contact the API server. However, if it regains contact, before it observes the gateway has transitioned, it could try to process globalnet resources and causes strange inconsistencies like we've seen. This is an edge case but perhaps we should stop the controllers if leadership is lost due to failure, especially since the controllers are likely offline as well. WDYT?

@tpantelis I believe that if the Lease cannot be updated due to API server issues, we can stop the controllers without deleting the globalnet chains (i.e., clearGlobalnetChains), as this action might disrupt the datapath. 
After the API server connectivity is restored, the initial Gateway can verify if it has transitioned to a non-Gateway node and then proceed to clearGlobalnetChains.

tpantelis committed:
Stop GN controllers when leadership is lost

If leadership is lost due to failure after the RenewDeadline expires,
currently we try to re-acquire leadership but we don't stop the controllers.
A failure can really only happen if the node loses contact with the API
server and can't update the Lease resource. However, it's possible there
could be a gateway transition during the time that leadership is lost and
a new instance is able to contact the API server to acquire the lock
(after the LeaseDuration expires). In this case, we'd have two controller
instances running although the first is most likely effectively disabled
anyway since most likely it can't contact the API server. However, if it
regains contact, before it observes the gateway has transitioned, it could
try to process globalnet resources and cause strange inconsistencies. This
is an edge case but is avoided by stopping the controllers if leadership is
lost due to failure, especially since the controllers are likely offline as
well. However we don't clear the globalnet chains to avoid dataplane
disruption.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented:
> @tpantelis I believe that if the Lease cannot be updated due to API server issues, we can stop the controllers without deleting the globalnet chains (i.e., clearGlobalnetChains), as this action might disrupt the datapath. After the API server connectivity is restored, the initial Gateway can verify if it has transitioned to a non-Gateway node and then proceed to clearGlobalnetChains.

OK - makes sense. Pushed a new [commit](https://github.com/submariner-io/submariner/pull/2743/commits/6ae6fc936d9913910caabb9011887f6db651dc07).

sridhargaddam commented:
@skitt can you please re-review?

sridhargaddam committed:
Merge branch 'devel' into gn_ctrl_race
